rfcs/0049-flakes.md

```diff
@@ -0,0 +1,708 @@
+flake references that don't specify an exact revision (e.g. `nixpkgs`
+rather than
+`github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082`). To
+ensure reproducibility, Nix will automatically generate and use a
+*lock file* called `flake.lock` in the flake's directory. The lock
+file contains a tree of mappings from the flake references specified
+in `flake.nix` to direct flake references that contain revisions.
+
+For example, if `flake.nix` contains
+```
+inputs.nixpkgs = {};
+inputs.import-cargo.uri = "github:edolstra/import-cargo";
+```
+then the resulting lock file might be:
+```
+{
+    "version": 3,
+    "inputs": {
+        "import-cargo": {
+            "inputs": {},
+            "narHash": "sha256-mxwKMDFOrhjrBQhIWwwm8mmEugyx/oVlvBH1CKxchlw=",
+            "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e",
+            "originalUri": "github:edolstra/import-cargo"
+        },
+        "nixpkgs": {
+            "inputs": {},
+            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
+            "uri": "github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082",
+            "originalUri": "nixpkgs"
+        }
+    }
+}
+```
+
+Thus, when we build this flake, the input `nixpkgs` is mapped to
+`github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e`. Nix
+will also check that the content hash of the input is equal to the one
+recorded in the lock file. This check is superfluous for Git
+repositories (since the commit hash serves a similar purpose), but for
+GitHub archives, we cannot directly check that the contents match the
+commit hash.
+
+Note that lock files are only used at top-level: the `flake.lock` files
+in dependencies (if they exist) are ignored.
+
+When you build a local repository (e.g. `nix build /path/to/repo`),
+Nix automatically creates a `flake.lock` file if it doesn't already
+exists. It will also update the lock file if inputs are added or
+removed. You can pass `--recreate-lock-file` to force Nix to recreate
+the lock file from scratch (and thus check for the latest version of
+each input).
+
+Lock files are not sufficient by themselves to ensure reproducible
+evaluation. It is also necessary to prevent certain impurities. In
+particular, the `nix` command now defaults to evaluating in "pure"
+mode, which means that the following are disallowed:
+
+* Access to files outside of the top-level flake or its inputs, as
+  well as paths fetched using `fetchTarball`, `fetchGit` and so on
+  without a commit hash or content hash. In particular this means that
+  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.
+
+* Access to the environment. This means that `builtins.getEnv "<var>"`
+  always returns an empty string.
+
+* Access to the system type (`builtins.currentSystem`).
+
+* Access to the current time (`builtins.currentTime`).
+
+* Use of the Nix search path (`<...>`); composition must be done
```

Comment:
What rationale is there for continuing the ```<...>``` support in pure mode? It seems like it is a commonly disliked feature of Nix that we have a chance to remove! I would rather this be disallowed completely and instead rely on ordinary function arguments in these cases.

Reply:
Maybe you misread that, because `<...>` *isn't* allowed in pure mode. (That was already the case before flakes.)